File: Iteration2/code/CSE3063GRP12/src/core/models/concretes/Semester.java
package core.models.concretes;

import java.util.Map;

import core.enums.CourseGrade;

public class Semester {
    private String id;
    private Map<String, CourseGrade> listOfCoursesTaken;
    private int creditsTaken;
    private int creditsCompleted;
    private double yano;
    private double gano;
    private int semesterNo;

    public Semester(String id, Map<String, CourseGrade> listOfCoursesTaken,
            int creditsTaken, double yano, int semesterNo) {
        this.id = id;
        this.listOfCoursesTaken = listOfCoursesTaken;
        this.creditsTaken = creditsTaken;
        this.yano = yano;
        this.semesterNo = semesterNo;
    }

    public String getId() {
        return id;
    }

    public Map<String, CourseGrade> getListOfCoursesTaken() {
        return listOfCoursesTaken;
    }

    public int getCreditsTaken() {
        return creditsTaken;
    }

    public int getCreditsCompleted() {
        return creditsCompleted;
    }

    public double getYano() {
        return yano;
    }

    public double getGano() {
        return gano;
    }

    public void setId(String id) {
        this.id = id;
    }

    public void setListOfCoursesTaken(Map<String, CourseGrade> listOfCoursesTaken) {
        this.listOfCoursesTaken = listOfCoursesTaken;
    }

    public void setCreditsTaken(int creditsTaken) {
        this.creditsTaken = creditsTaken;
    }

    public void setCreditsCompleted(int creditsCompleted) {
        this.creditsCompleted = creditsCompleted;
    }

    public void setYano(double yano) {
        this.yano = yano;
    }

    public void setGano(double gano) {
        this.gano = gano;
    }

    public int getSemesterNo() {
        return semesterNo;
    }

    public void setSemesterNo(int semesterNo) {
        this.semesterNo = semesterNo;
    }

}

File: Iteration3/code/CSE3063GRP12/src/core/models/concretes/Semester.py
class Semester:
    def __init__(self, id, listOfCoursesTaken, creditsTaken, yano, semesterNo):
        self.id = id
        self.listOfCoursesTaken = listOfCoursesTaken  # dictionary mapping course identifier to CourseGrade
        self.credits_taken = creditsTaken
        self.yano = yano 
        self.semesterNo = semesterNo
    
    def to_dict(self):
        return {
            "id": self.id,
            "listOfCoursesTaken": {courseId: courseGrade.to_dict() if hasattr(courseGrade, 'to_dict') else courseGrade for courseId, courseGrade in self.listOfCoursesTaken.items()},
            "credits_taken": self.credits_taken,
            "yano": self.yano,
            "semesterNo": self.semesterNo
        }